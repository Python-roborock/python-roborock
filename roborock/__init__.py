"""Roborock API.

.. include:: ../README.md
"""

from roborock.b01_containers import *
from roborock.code_mappings import *
from roborock.containers import *
from roborock.exceptions import *
from roborock.roborock_typing import *

__all__ = [
    "web_api",
    "version_1_apis",
    "version_a01_apis",
    "containers",
    "b01_containers",
    "const",
    "cloud_api",
    "clean_modes",
    "code_mappings",
    "roborock_typing",
    "exceptions",

    # We'll add new APIs here in the future e.g. devices/
]
