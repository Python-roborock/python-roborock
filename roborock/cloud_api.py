from __future__ import annotations

import logging
import threading
import uuid
from abc import ABC
from asyncio import Lock
from typing import Any
from urllib.parse import urlparse

import paho.mqtt.client as mqtt

from .api import KEEPALIVE, RoborockClient
from .containers import DeviceData, UserData
from .exceptions import RoborockException, VacuumError
from .protocol import MessageParser, md5hex
from .roborock_future import RoborockFuture

_LOGGER = logging.getLogger(__name__)
CONNECT_REQUEST_ID = 0
DISCONNECT_REQUEST_ID = 1


class RoborockMqttClient(RoborockClient, mqtt.Client, ABC):
    _thread: threading.Thread
    _client_id: str

    def __init__(self, user_data: UserData, device_info: DeviceData, queue_timeout: int = 10) -> None:
        rriot = user_data.rriot
        if rriot is None:
            raise RoborockException("Got no rriot data from user_data")
        RoborockClient.__init__(self, device_info, queue_timeout)
        mqtt.Client.__init__(self, protocol=mqtt.MQTTv5)
        self._mqtt_user = rriot.u
        self._hashed_user = md5hex(self._mqtt_user + ":" + rriot.k)[2:10]
        url = urlparse(rriot.r.m)
        if not isinstance(url.hostname, str):
            raise RoborockException("Url parsing returned an invalid hostname")
        self._mqtt_host = str(url.hostname)
        self._mqtt_port = url.port
        self._mqtt_ssl = url.scheme == "ssl"
        if self._mqtt_ssl:
            super().tls_set()
        self._mqtt_password = rriot.s
        self._hashed_password = md5hex(self._mqtt_password + ":" + rriot.k)[16:]
        super().username_pw_set(self._hashed_user, self._hashed_password)
        self._waiting_queue: dict[int, RoborockFuture] = {}
        self._mutex = Lock()
        self.update_client_id()

    def on_connect(self, *args, **kwargs):
        _, __, ___, rc, ____ = args
        connection_queue = self._waiting_queue.get(CONNECT_REQUEST_ID)
        if rc != mqtt.MQTT_ERR_SUCCESS:
            message = f"Failed to connect ({mqtt.error_string(rc)})"
            self._logger.error(message)
            if connection_queue:
                connection_queue.set_exception(VacuumError(message))
            return
        self._logger.info(f"Connected to mqtt {self._mqtt_host}:{self._mqtt_port}")
        topic = f"rr/m/o/{self._mqtt_user}/{self._hashed_user}/{self.device_info.device.duid}"
        (result, mid) = self.subscribe(topic)
        if result != 0:
            message = f"Failed to subscribe ({mqtt.error_string(rc)})"
            self._logger.error(message)
            if connection_queue:
                connection_queue.set_exception(VacuumError(message))
            return
        self._logger.info(f"Subscribed to topic {topic}")
        if connection_queue:
            connection_queue.set_result(True)

    def on_message(self, *args, **kwargs):
        client, __, msg = args
        try:
            messages, _ = MessageParser.parse(msg.payload, local_key=self.device_info.device.local_key)
            super().on_message_received(messages)
        except Exception as ex:
            self._logger.exception(ex)

    def on_disconnect(self, *args, **kwargs):
        _, __, rc, ___ = args
        try:
            exc = RoborockException(mqtt.error_string(rc)) if rc != mqtt.MQTT_ERR_SUCCESS else None
            super().on_connection_lost(exc)
            if rc == mqtt.MQTT_ERR_PROTOCOL:
                self.update_client_id()
            connection_queue = self._waiting_queue.get(DISCONNECT_REQUEST_ID)
            if connection_queue:
                connection_queue.set_result(True)
        except Exception as ex:
            self._logger.exception(ex)

    def update_client_id(self):
        self._client_id = mqtt.base62(uuid.uuid4().int, padding=22)

    def sync_stop_loop(self) -> None:
        if self._thread:
            self._logger.info("Stopping mqtt loop")
            super().loop_stop()

    def sync_start_loop(self) -> None:
        if not self._thread or not self._thread.is_alive():
            self.sync_stop_loop()
            self._logger.info("Starting mqtt loop")
            super().loop_start()

    def sync_disconnect(self) -> Any:
        if not self.is_connected():
            return None

        self._logger.info("Disconnecting from mqtt")
        disconnected_future = self._async_response(DISCONNECT_REQUEST_ID)
        rc = super().disconnect()

        if rc == mqtt.MQTT_ERR_NO_CONN:
            disconnected_future.cancel()
            return None

        if rc != mqtt.MQTT_ERR_SUCCESS:
            disconnected_future.cancel()
            raise RoborockException(f"Failed to disconnect ({mqtt.error_string(rc)})")

        return disconnected_future

    def sync_connect(self) -> Any:
        if self.is_connected():
            self.sync_start_loop()
            return None

        if self._mqtt_port is None or self._mqtt_host is None:
            raise RoborockException("Mqtt information was not entered. Cannot connect.")

        self._logger.debug("Connecting to mqtt")
        connected_future = self._async_response(CONNECT_REQUEST_ID)
        super().connect(host=self._mqtt_host, port=self._mqtt_port, keepalive=KEEPALIVE)

        self.sync_start_loop()
        return connected_future

    async def async_disconnect(self) -> None:
        async with self._mutex:
            if disconnected_future := self.sync_disconnect():
                try:
                    await disconnected_future
                except VacuumError as err:
                    raise RoborockException(err) from err

    async def async_connect(self) -> None:
        async with self._mutex:
            if connected_future := self.sync_connect():
                try:
                    await connected_future
                except VacuumError as err:
                    raise RoborockException(err) from err

    def _send_msg_raw(self, msg: bytes) -> None:
        info = self.publish(f"rr/m/i/{self._mqtt_user}/{self._hashed_user}/{self.device_info.device.duid}", msg)
        if info.rc != mqtt.MQTT_ERR_SUCCESS:
            raise RoborockException(f"Failed to publish ({mqtt.error_string(info.rc)})")
