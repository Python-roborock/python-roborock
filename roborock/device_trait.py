from abc import ABC, abstractmethod
from collections.abc import Awaitable, Callable
from dataclasses import dataclass

from . import RoborockCommand
from .containers import DeviceFeatures


@dataclass
class DeviceTrait(ABC):
    handle_command: RoborockCommand

    def __init__(self, send_command: Callable[..., Awaitable[None]]):
        self.send_command = send_command
        self.subscriptions = []

    @classmethod
    @abstractmethod
    def supported(cls, features: DeviceFeatures) -> bool:
        raise NotImplementedError

    @abstractmethod
    def update(cls, data: dict) -> bool:
        raise NotImplementedError

    def on_message(self, data: dict) -> None:
        self.status = self.update(data)
        for callback in self.subscriptions:
            callback(self.status)

    def subscribe(self, callable: Callable):
        # Maybe needs to handle async too?
        self.subscriptions.append(callable)

    @abstractmethod
    def get(self):
        raise NotImplementedError


# class ConsumableTrait(DeviceTrait):
#     handle_command = RoborockCommand.GET_CONSUMABLE
#     _status_type: type[Consumable] = DnDTimer
#     status: Consumable
#
#     def __init__(self, send_command: Callable[..., Awaitable[None]]):
#         super().__init__(send_command)
#
#     @classmethod
#     def supported(cls, features: DeviceFeatures) -> bool:
#         return True
#
#     async def reset_consumable(self, consumable: str) -> None:
#         await self.send_command(RoborockCommand.RESET_CONSUMABLE, [consumable])
#
#     async def get(self) -> None:
#         await self.send_command(RoborockCommand.GET_CONSUMABLE)
