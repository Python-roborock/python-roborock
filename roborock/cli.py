from __future__ import annotations

import asyncio
import json
import logging
from dataclasses import asdict
from pathlib import Path
from typing import Any

import click
from pyshark import FileCapture  # type: ignore
from pyshark.capture.live_capture import LiveCapture, UnknownInterfaceException  # type: ignore
from pyshark.packet.packet import Packet  # type: ignore

from roborock import SHORT_MODEL_TO_ENUM, DeviceFeatures, RoborockCommand, RoborockException
from roborock.containers import DeviceData, HomeData, HomeDataProduct, LoginData
from roborock.devices.device_manager import create_device_manager, create_home_data_api
from roborock.protocol import MessageParser
from roborock.util import run_sync
from roborock.version_1_apis.roborock_local_client_v1 import RoborockLocalClientV1
from roborock.version_1_apis.roborock_mqtt_client_v1 import RoborockMqttClientV1
from roborock.web_api import RoborockApiClient

_LOGGER = logging.getLogger(__name__)


class RoborockContext:
    roborock_file = Path("~/.roborock").expanduser()
    _login_data: LoginData | None = None

    def __init__(self):
        self.reload()

    def reload(self):
        if self.roborock_file.is_file():
            with open(self.roborock_file) as f:
                data = json.load(f)
                if data:
                    self._login_data = LoginData.from_dict(data)

    def update(self, login_data: LoginData):
        data = json.dumps(login_data.as_dict(), default=vars)
        with open(self.roborock_file, "w") as f:
            f.write(data)
        self.reload()

    def validate(self):
        if self._login_data is None:
            raise RoborockException("You must login first")

    def login_data(self) -> LoginData:
        """Get the login data."""
        self.validate()
        return self._login_data


@click.option("-d", "--debug", default=False, count=True)
@click.version_option(package_name="python-roborock")
@click.group()
@click.pass_context
def cli(ctx, debug: int):
    logging_config: dict[str, Any] = {"level": logging.DEBUG if debug > 0 else logging.INFO}
    logging.basicConfig(**logging_config)  # type: ignore
    ctx.obj = RoborockContext()


@click.command()
@click.option("--email", required=True)
@click.option(
    "--password",
    required=False,
    help="Password for the Roborock account. If not provided, an email code will be requested.",
)
@click.pass_context
@run_sync()
async def login(ctx, email, password):
    """Login to Roborock account."""
    context: RoborockContext = ctx.obj
    try:
        context.validate()
        _LOGGER.info("Already logged in")
        return
    except RoborockException:
        pass
    client = RoborockApiClient(email)
    if password is not None:
        user_data = await client.pass_login(password)
    else:
        print(f"Requesting code for {email}")
        await client.request_code()
        code = click.prompt("A code has been sent to your email, please enter the code", type=str)
        user_data = await client.code_login(code)
        print("Login successful")
    context.update(LoginData(user_data=user_data, email=email))


@click.command()
@click.pass_context
@click.option("--duration", default=10, help="Duration to run the MQTT session in seconds")
@run_sync()
async def session(ctx, duration: int):
    context: RoborockContext = ctx.obj
    login_data = context.login_data()

    home_data_api = create_home_data_api(login_data.email, login_data.user_data)

    async def home_data_cache() -> HomeData:
        if login_data.home_data is None:
            login_data.home_data = await home_data_api()
            context.update(login_data)
        return login_data.home_data

    # Create device manager
    device_manager = await create_device_manager(login_data.user_data, home_data_cache)

    devices = await device_manager.get_devices()
    click.echo(f"Discovered devices: {', '.join([device.name for device in devices])}")

    click.echo("MQTT session started. Querying devices...")
    for device in devices:
        if not (status_trait := device.traits.get("status")):
            click.echo(f"Device {device.name} does not have a status trait")
            continue
        try:
            status = await status_trait.get_status()
        except RoborockException as e:
            click.echo(f"Failed to get status for {device.name}: {e}")
        else:
            click.echo(f"Device {device.name} status: {status.as_dict()}")

    click.echo("Listening for messages.")
    await asyncio.sleep(duration)

    # Close the device manager (this will close all devices and MQTT session)
    await device_manager.close()


async def _discover(ctx):
    context: RoborockContext = ctx.obj
    login_data = context.login_data()
    if not login_data:
        raise Exception("You need to login first")
    client = RoborockApiClient(login_data.email)
    home_data = await client.get_home_data(login_data.user_data)
    login_data.home_data = home_data
    context.update(login_data)
    click.echo(f"Discovered devices {', '.join([device.name for device in home_data.get_all_devices()])}")


@click.command()
@click.pass_context
@run_sync()
async def discover(ctx):
    await _discover(ctx)


@click.command()
@click.pass_context
@run_sync()
async def list_devices(ctx):
    context: RoborockContext = ctx.obj
    login_data = context.login_data()
    if not login_data.home_data:
        await _discover(ctx)
        login_data = context.login_data()
    home_data = login_data.home_data
    device_name_id = ", ".join(
        [f"{device.name}: {device.duid}" for device in home_data.devices + home_data.received_devices]
    )
    click.echo(f"Known devices {device_name_id}")


@click.command()
@click.option("--device_id", required=True)
@click.pass_context
@run_sync()
async def list_scenes(ctx, device_id):
    context: RoborockContext = ctx.obj
    login_data = context.login_data()
    if not login_data.home_data:
        await _discover(ctx)
        login_data = context.login_data()
    client = RoborockApiClient(login_data.email)
    scenes = await client.get_scenes(login_data.user_data, device_id)
    output_list = []
    for scene in scenes:
        output_list.append(scene.as_dict())
    click.echo(json.dumps(output_list, indent=4))


@click.command()
@click.option("--scene_id", required=True)
@click.pass_context
@run_sync()
async def execute_scene(ctx, scene_id):
    context: RoborockContext = ctx.obj
    login_data = context.login_data()
    if not login_data.home_data:
        await _discover(ctx)
        login_data = context.login_data()
    client = RoborockApiClient(login_data.email)
    await client.execute_scene(login_data.user_data, scene_id)


@click.command()
@click.option("--device_id", required=True)
@click.pass_context
@run_sync()
async def status(ctx, device_id):
    context: RoborockContext = ctx.obj
    login_data = context.login_data()
    if not login_data.home_data:
        await _discover(ctx)
        login_data = context.login_data()
    home_data = login_data.home_data
    devices = home_data.devices + home_data.received_devices
    device = next(device for device in devices if device.duid == device_id)
    product_info: dict[str, HomeDataProduct] = {product.id: product for product in home_data.products}
    device_data = DeviceData(device, product_info[device.product_id].model)
    mqtt_client = RoborockMqttClientV1(login_data.user_data, device_data)
    networking = await mqtt_client.get_networking()
    local_device_data = DeviceData(device, product_info[device.product_id].model, networking.ip)
    local_client = RoborockLocalClientV1(local_device_data)
    status = await local_client.get_status()
    click.echo(json.dumps(status.as_dict(), indent=4))


@click.command()
@click.option("--device_id", required=True)
@click.option("--cmd", required=True)
@click.option("--params", required=False)
@click.pass_context
@run_sync()
async def command(ctx, cmd, device_id, params):
    context: RoborockContext = ctx.obj
    login_data = context.login_data()
    if not login_data.home_data:
        await _discover(ctx)
        login_data = context.login_data()
    home_data = login_data.home_data
    devices = home_data.devices + home_data.received_devices
    device = next(device for device in devices if device.duid == device_id)
    model = next(
        (product.model for product in home_data.products if device is not None and product.id == device.product_id),
        None,
    )
    if model is None:
        raise RoborockException(f"Could not find model for device {device.name}")
    device_info = DeviceData(device=device, model=model)
    mqtt_client = RoborockMqttClientV1(login_data.user_data, device_info)
    await mqtt_client.send_command(cmd, json.loads(params) if params is not None else None)
    await mqtt_client.async_release()


@click.command()
@click.option("--local_key", required=True)
@click.option("--device_ip", required=True)
@click.option("--file", required=False)
@click.pass_context
@run_sync()
async def parser(_, local_key, device_ip, file):
    file_provided = file is not None
    if file_provided:
        capture = FileCapture(file)
    else:
        _LOGGER.info("Listen for interface rvi0 since no file was provided")
        capture = LiveCapture(interface="rvi0")
    buffer = {"data": b""}

    def on_package(packet: Packet):
        if hasattr(packet, "ip"):
            if packet.transport_layer == "TCP" and (packet.ip.dst == device_ip or packet.ip.src == device_ip):
                if hasattr(packet, "DATA"):
                    if hasattr(packet.DATA, "data"):
                        if packet.ip.dst == device_ip:
                            try:
                                f, buffer["data"] = MessageParser.parse(
                                    buffer["data"] + bytes.fromhex(packet.DATA.data),
                                    local_key,
                                )
                                print(f"Received request: {f}")
                            except BaseException as e:
                                print(e)
                                pass
                        elif packet.ip.src == device_ip:
                            try:
                                f, buffer["data"] = MessageParser.parse(
                                    buffer["data"] + bytes.fromhex(packet.DATA.data),
                                    local_key,
                                )
                                print(f"Received response: {f}")
                            except BaseException as e:
                                print(e)
                                pass

    try:
        await capture.packets_from_tshark(on_package, close_tshark=not file_provided)
    except UnknownInterfaceException:
        raise RoborockException(
            "You need to run 'rvictl -s XXXXXXXX-XXXXXXXXXXXXXXXX' first, with an iPhone connected to usb port"
        )


@click.command()
@click.pass_context
@run_sync()
async def update_docs(ctx):
    """
    Generates or updates a markdown file with features for all devices.
    """
    MARKDOWN_FILE = Path("../SUPPORTED_FEATURES.md")

    def write_markdown_table(product_data: dict[str, dict[str, str]], all_features: set[str]):
        """Writes the data into a markdown table (products as columns)."""
        sorted_products = sorted(product_data.keys())
        sorted_features = sorted(list(all_features))

        header = ["Feature"] + sorted_products

        with open(MARKDOWN_FILE, "w", encoding="utf-8") as f:
            f.write("| " + " | ".join(header) + " |\n")
            f.write("|" + "---|" * len(header) + "\n")

            # Create data for special rows
            special_rows = [
                "Product Nickname",
                "Protocol Version",
                "New Feature Info",
                "New Feature Info Str",
                "Feature Info",
            ]
            for row in special_rows:
                row_values = [str(product_data[p].get(row, "")) for p in sorted_products]
                f.write("| " + " | ".join([row] + row_values) + " |\n")
            for feature in sorted_features:
                feature_row = [f"`{feature}`"]
                for product in sorted_products:
                    feature_row.append(product_data[product].get(feature, ""))
                f.write("| " + " | ".join(feature_row) + " |\n")

    product_features_map = {}
    all_feature_names = set()

    context: RoborockContext = ctx.obj
    login_data = context.login_data()
    if not login_data.home_data:
        await _discover(ctx)
        login_data = context.login_data()
    home_data = login_data.home_data

    all_devices = home_data.devices + home_data.received_devices
    click.echo(f"Found {len(all_devices)} devices. Fetching current data via MQTT...")

    for device in all_devices:
        click.echo(f"  - Processing {device.name} ({device.duid})")
        product_info = home_data.product_map[device.product_id]
        device_data = DeviceData(device, product_info.model)
        mqtt_client = RoborockMqttClientV1(login_data.user_data, device_data)
        try:
            init_status_result = await mqtt_client.send_command(
                RoborockCommand.APP_GET_INIT_STATUS,
            )
            product_nickname = SHORT_MODEL_TO_ENUM.get(product_info.model.split(".")[-1])
            device_features = DeviceFeatures.from_feature_flags(
                new_feature_info=init_status_result.get("new_feature_info"),
                new_feature_info_str=init_status_result.get("new_feature_info_str"),
                feature_info=init_status_result.get("feature_info"),
                product_nickname=product_nickname,
            )
            features_dict = asdict(device_features)

            current_product_data = {
                "Protocol Version": device.pv,
                "Product Nickname": product_nickname,
                "New Feature Info": init_status_result.get("new_feature_info"),
                "New Feature Info Str": init_status_result.get("new_feature_info_str"),
                "Feature Info": init_status_result.get("feature_info"),
            }

            for feature, is_supported in features_dict.items():
                all_feature_names.add(feature)
                current_product_data[feature] = "X" if is_supported else ""

            product_features_map[product_info.model] = current_product_data

        except Exception as e:
            click.echo(f"    - Error processing device {device.name}: {e}")
        finally:
            await mqtt_client.async_release()

    if not product_features_map:
        click.echo("No device data could be gathered. File not updated.")
        return

    click.echo(f"Writing updated data to {MARKDOWN_FILE}...")
    write_markdown_table(product_features_map, all_feature_names)
    click.echo("Done.")


cli.add_command(login)
cli.add_command(discover)
cli.add_command(list_devices)
cli.add_command(list_scenes)
cli.add_command(execute_scene)
cli.add_command(status)
cli.add_command(command)
cli.add_command(parser)
cli.add_command(session)
cli.add_command(update_docs)


def main():
    return cli()


if __name__ == "__main__":
    main()
