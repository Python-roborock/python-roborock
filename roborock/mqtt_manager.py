from __future__ import annotations

import asyncio
import dataclasses
import logging
from collections.abc import Coroutine
from typing import Callable, Self
from urllib.parse import urlparse

import aiomqtt
from aiomqtt import TLSParameters

from roborock import RoborockException, UserData
from roborock.protocol import MessageParser, md5hex

from .containers import DeviceData

LOGGER = logging.getLogger(__name__)


@dataclasses.dataclass
class ClientWrapper:
    publish_function: Coroutine[None]
    unsubscribe_function: Coroutine[None]
    subscribe_function: Coroutine[None]


class RoborockMqttManager:
    client_wrappers: dict[str, ClientWrapper] = {}
    _instance: Self = None

    def __new__(cls) -> RoborockMqttManager:
        if cls._instance is None:
            cls._instance = super().__new__(cls)
        return cls._instance

    async def connect(self, user_data: UserData):
        # Add some kind of lock so we don't try to connect if we are already trying to connect the same account.
        if user_data.rriot.u not in self.client_wrappers:
            loop = asyncio.get_event_loop()
            loop.create_task(self._new_connect(user_data))

    async def _new_connect(self, user_data: UserData):
        rriot = user_data.rriot
        mqtt_user = rriot.u
        hashed_user = md5hex(mqtt_user + ":" + rriot.k)[2:10]
        url = urlparse(rriot.r.m)
        if not isinstance(url.hostname, str):
            raise RoborockException("Url parsing returned an invalid hostname")
        mqtt_host = str(url.hostname)
        mqtt_port = url.port

        mqtt_password = rriot.s
        hashed_password = md5hex(mqtt_password + ":" + rriot.k)[16:]
        LOGGER.debug("Connecting to %s for %s", mqtt_host, mqtt_user)

        async with aiomqtt.Client(
            hostname=mqtt_host,
            port=mqtt_port,
            username=hashed_user,
            password=hashed_password,
            keepalive=60,
            tls_params=TLSParameters(),
        ) as client:
            # TODO: Handle logic for when client loses connection
            LOGGER.info("Connected to %s for %s", mqtt_host, mqtt_user)
            callbacks: dict[str, Callable] = {}
            device_map = {}

            async def publish(device: DeviceData, payload: bytes):
                await client.publish(f"rr/m/i/{mqtt_user}/{hashed_user}/{device.device.duid}", payload=payload)

            async def subscribe(device: DeviceData, callback):
                LOGGER.debug(f"Subscribing to rr/m/o/{mqtt_user}/{hashed_user}/{device.device.duid}")
                await client.subscribe(f"rr/m/o/{mqtt_user}/{hashed_user}/{device.device.duid}")
                LOGGER.debug(f"Subscribed to rr/m/o/{mqtt_user}/{hashed_user}/{device.device.duid}")
                callbacks[device.device.duid] = callback
                device_map[device.device.duid] = device
                return

            async def unsubscribe(device: DeviceData):
                await client.unsubscribe(f"rr/m/o/{mqtt_user}/{hashed_user}/{device.device.duid}")

            self.client_wrappers[user_data.rriot.u] = ClientWrapper(
                publish_function=publish, unsubscribe_function=unsubscribe, subscribe_function=subscribe
            )
            async for message in client.messages:
                try:
                    device_id = message.topic.value.split("/")[-1]
                    device = device_map[device_id]
                    message = MessageParser.parse(message.payload, device.device.local_key)
                    for m in message[0]:
                        callbacks[device_id](m)
                except Exception:
                    ...

    async def disconnect(self, user_data: UserData):
        await self.client_wrappers[user_data.rriot.u].disconnect()

    async def subscribe(self, user_data: UserData, device: DeviceData, callback):
        if user_data.rriot.u not in self.client_wrappers:
            await self.connect(user_data)
        # add some kind of lock to make sure we don't subscribe until the connection is successful
        await asyncio.sleep(2)
        await self.client_wrappers[user_data.rriot.u].subscribe_function(device, callback)

    async def unsubscribe(self):
        pass

    async def publish(self, user_data: UserData, device, payload: bytes):
        LOGGER.debug("Publishing topic for %s, Message: %s", device.device.duid, payload)
        if user_data.rriot.u not in self.client_wrappers:
            await self.connect(user_data)
        await self.client_wrappers[user_data.rriot.u].publish_function(device, payload)
