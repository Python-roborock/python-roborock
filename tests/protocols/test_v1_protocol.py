"""Tests for the v1 protocol message encoding and decoding."""

from collections.abc import Generator
from unittest.mock import patch

import pytest
from freezegun import freeze_time

from roborock.containers import RoborockBase, UserData
from roborock.protocols.v1_protocol import (
    SecurityData,
    create_mqtt_payload_encoder,
    decode_rpc_response,
    encode_local_payload,
)
from roborock.roborock_message import RoborockMessage, RoborockMessageProtocol
from roborock.roborock_typing import RoborockCommand

from .. import mock_data

USER_DATA = UserData.from_dict(mock_data.USER_DATA)
TEST_REQUEST_ID = 44444
SECURITY_DATA = SecurityData(endpoint="3PBTIjvc", nonce=b"fake-nonce")


@pytest.fixture(autouse=True)
def fixed_time_fixture() -> Generator[None, None, None]:
    """Fixture to freeze time for predictable request IDs."""
    # Freeze time to a specific point so request IDs are predictable
    with freeze_time("2025-01-20T12:00:00"):
        yield


@pytest.fixture(name="test_request_id", autouse=True)
def request_id_fixture() -> Generator[int, None, None]:
    """Fixture to provide a fixed request ID."""
    with patch("roborock.protocols.v1_protocol.get_next_int", return_value=TEST_REQUEST_ID):
        yield TEST_REQUEST_ID


@pytest.mark.parametrize(
    ("command", "params", "expected"),
    [
        (
            RoborockCommand.GET_STATUS,
            None,
            b'{"dps":{"101":"{\\"id\\":44444,\\"method\\":\\"get_status\\",\\"params\\":[]}"},"t":1737374400}',
        )
    ],
)
def test_encode_local_payload(command, params, expected):
    """Test encoding of local payload for V1 commands."""
    message = encode_local_payload(command, params)
    assert isinstance(message, RoborockMessage)
    assert message.protocol == RoborockMessageProtocol.GENERAL_REQUEST
    assert message.payload == expected


@pytest.mark.parametrize(
    ("command", "params", "expected"),
    [
        (
            RoborockCommand.GET_STATUS,
            None,
            b'{"dps":{"101":"{\\"id\\":44444,\\"method\\":\\"get_status\\",\\"params\\":[],\\"security\\":{\\"endpoint\\":\\"3PBTIjvc\\",\\"nonce\\":\\"66616b652d6e6f6e6365\\"}}"},"t":1737374400}',
        )
    ],
)
def test_encode_mqtt_payload(command, params, expected):
    """Test encoding of local payload for V1 commands."""
    encoder = create_mqtt_payload_encoder(SECURITY_DATA)
    message = encoder(command, params)
    assert isinstance(message, RoborockMessage)
    assert message.protocol == RoborockMessageProtocol.RPC_REQUEST
    assert message.payload == expected


@pytest.mark.parametrize(
    ("payload", "expected"),
    [
        (
            b'{"t":1652547161,"dps":{"102":"{\\"id\\":20005,\\"result\\":[{\\"msg_ver\\":2,\\"msg_seq\\":1072,\\"state\\":8,\\"battery\\":100,\\"clean_time\\":1041,\\"clean_area\\":37080000,\\"error_code\\":0,\\"map_present\\":1,\\"in_cleaning\\":0,\\"in_returning\\":0,\\"in_fresh_state\\":1,\\"lab_status\\":1,\\"water_box_status\\":0,\\"fan_power\\":103,\\"dnd_enabled\\":0,\\"map_status\\":3,\\"is_locating\\":0,\\"lock_status\\":0,\\"water_box_mode\\":202,\\"distance_off\\":0,\\"water_box_carriage_status\\":0,\\"mop_forbidden_enable\\":0,\\"unsave_map_reason\\":0,\\"unsave_map_flag\\":0}]}"}}',
            {
                "msg_ver": 2,
                "msg_seq": 1072,
                "state": 8,
                "battery": 100,
                "clean_time": 1041,
                "clean_area": 37080000,
                "error_code": 0,
                "map_present": 1,
                "in_cleaning": 0,
                "in_returning": 0,
                "in_fresh_state": 1,
                "lab_status": 1,
                "water_box_status": 0,
                "fan_power": 103,
                "dnd_enabled": 0,
                "map_status": 3,
                "is_locating": 0,
                "lock_status": 0,
                "water_box_mode": 202,
                "distance_off": 0,
                "water_box_carriage_status": 0,
                "mop_forbidden_enable": 0,
                "unsave_map_reason": 0,
                "unsave_map_flag": 0,
            },
        ),
    ],
)
def test_decode_rpc_response(payload: bytes, expected: RoborockBase) -> None:
    """Test decoding a v1 RPC response protocol message."""
    # The values other than the payload are arbitrary
    message = RoborockMessage(
        protocol=RoborockMessageProtocol.GENERAL_RESPONSE,
        payload=payload,
        seq=12750,
        version=b"1.0",
        random=97431,
        timestamp=1652547161,
    )
    decoded_message = decode_rpc_response(message)
    assert decoded_message == expected
