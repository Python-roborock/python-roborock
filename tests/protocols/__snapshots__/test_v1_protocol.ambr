# serializer version: 1
# name: test_decode_rpc_payload[get_clean_summary]
  20001
# ---
# name: test_decode_rpc_payload[get_clean_summary].1
  '''
  [
    1442559,
    24258125000,
    296,
    [
      1756848207,
      1754930385,
      1753203976,
      1752183435,
      1747427370,
      1746204046,
      1745601543,
      1744387080,
      1743528522,
      1742489154,
      1741022299,
      1740433682,
      1739902516,
      1738875106,
      1738864366,
      1738620067,
      1736873889,
      1736197544,
      1736121269,
      1734458038
    ]
  ]
  '''
# ---
# name: test_decode_rpc_payload[get_dnd]
  20002
# ---
# name: test_decode_rpc_payload[get_dnd].1
  '''
  [
    {
      "start_hour": 22,
      "start_minute": 0,
      "end_hour": 8,
      "end_minute": 0,
      "enabled": 1
    }
  ]
  '''
# ---
# name: test_decode_rpc_payload[get_status]
  20001
# ---
# name: test_decode_rpc_payload[get_status].1
  '''
  [
    {
      "msg_ver": 2,
      "msg_seq": 515,
      "state": 8,
      "battery": 100,
      "clean_time": 5405,
      "clean_area": 91287500,
      "error_code": 0,
      "map_present": 1,
      "in_cleaning": 0,
      "in_returning": 0,
      "in_fresh_state": 1,
      "lab_status": 1,
      "water_box_status": 0,
      "fan_power": 106,
      "dnd_enabled": 1,
      "map_status": 3,
      "is_locating": 0,
      "lock_status": 0,
      "water_box_mode": 204,
      "distance_off": 0,
      "water_box_carriage_status": 0,
      "mop_forbidden_enable": 0,
      "unsave_map_reason": 4,
      "unsave_map_flag": 0
    }
  ]
  '''
# ---
